cpp/src/parquet/chunker_internal.cc

```diff
@@ -0,0 +1,413 @@
+      DCHECK_GE(chunk.value_offset, prev_chunk.value_offset);
+      DCHECK_EQ(chunk.level_offset, prev_chunk.level_offset + prev_chunk.levels_to_write);
+      sum_levels += chunk.levels_to_write;
+    }
+    DCHECK_EQ(sum_levels, num_levels);
+
+    // the last chunk must end at the last level
+    auto last_chunk = chunks.back();
+    DCHECK_EQ(last_chunk.level_offset + last_chunk.levels_to_write, num_levels);
+  }
+
+  template <typename RollFunc>
+  std::vector<Chunk> Calculate(const int16_t* def_levels, const int16_t* rep_levels,
+                               int64_t num_levels, const RollFunc& RollValue) {
+    // Calculate the chunk boundaries for typed Arrow arrays.
+    std::vector<Chunk> chunks;
+    int64_t offset;
+    int64_t prev_offset = 0;
+    int64_t prev_value_offset = 0;
+    bool has_def_levels = level_info_.def_level > 0;
+    bool has_rep_levels = level_info_.rep_level > 0;
+
+    if (!has_rep_levels && !has_def_levels) {
+      // fastest path for non-nested non-null data
+      for (offset = 0; offset < num_levels; ++offset) {
+        RollValue(offset);
+        if (NeedNewChunk()) {
+          chunks.push_back({prev_offset, prev_offset, offset - prev_offset});
+          prev_offset = offset;
+        }
+      }
+      // set the previous value offset to add the last chunk
+      prev_value_offset = prev_offset;
+    } else if (!has_rep_levels) {
+      // non-nested data with nulls
+      int16_t def_level;
+      for (int64_t offset = 0; offset < num_levels; ++offset) {
+        def_level = def_levels[offset];
+
+        Roll(&def_level);
+        if (def_level == level_info_.def_level) {
+          RollValue(offset);
+        }
+        if (NeedNewChunk()) {
+          chunks.push_back({prev_offset, prev_offset, offset - prev_offset});
+          prev_offset = offset;
+        }
+      }
+      // set the previous value offset to add the last chunk
+      prev_value_offset = prev_offset;
+    } else {
+      // nested data with nulls
+      int16_t def_level;
+      int16_t rep_level;
+      int64_t value_offset = 0;
+
+      for (offset = 0; offset < num_levels; ++offset) {
+        def_level = def_levels[offset];
+        rep_level = rep_levels[offset];
+
+        Roll(&def_level);
+        Roll(&rep_level);
+        if (def_level == level_info_.def_level) {
+          RollValue(value_offset);
+        }
+
+        if ((rep_level == 0) && NeedNewChunk()) {
+          // if we are at a record boundary and need a new chunk, we create a new chunk
+          auto levels_to_write = offset - prev_offset;
+          if (levels_to_write > 0) {
+            chunks.push_back({prev_offset, prev_value_offset, levels_to_write});
+            prev_offset = offset;
+            prev_value_offset = value_offset;
+          }
+        }
+        if (def_level >= level_info_.repeated_ancestor_def_level) {
+          // we only increment the value offset if we have a leaf value
+          ++value_offset;
+        }
+      }
```

Comment:
Do we need the following that exists in other clauses here too?

```cpp
      // set the previous value offset to add the last chunk
      prev_value_offset = prev_offset;
```

Reply:
No, the other two branches are not incrementing `prev_value_offset` separately to the level offsets se we set `prev_value_offset = prev_[level]_offset` in those branches. The last branch handles the nested case and it maintains `prev_value_offset` already.

Reply:
This way I was able to have a single 

```
    // add the last chunk if we have any levels left
    if (prev_offset < num_levels) {
      chunks.push_back({prev_offset, prev_value_offset, num_levels - prev_offset});
    }
```

at the end of the function rather than at the and of all three branches.